### PR TITLE
[DNM] Deterministic hashing for almost everything

### PR DESCRIPTION
**Pull request by crusaderky**
- Builds on top of #10876 
- Alternative to #10808

### CODE REVIEW

**Review by milesgranger — APPROVED** [review]
Nothing stands out to me, looks good. Thanks, @crusaderky!

File: dask/base.py [hunk 1]
```diff
@@ -1147,105 +1147,92 @@ def normalize_random_state(state):
...
+    try:
+        pik = pickle.dumps(o, protocol=5, buffer_callback=buffers.append)
+        if b"__main__" in pik:
+            pik = None
+    except Exception:
+        pik = None
+
+    if pik is None:
+        buffers.clear()
+        pik = cloudpickle.dumps(o, protocol=5, buffer_callback=buffers.append)
```
Comment (fjetter) [inline dask/base.py:1222]: `main` disallows using cloudpickle when `ensure-deterministic` is enabled. This behavior was introduced in https://github.com/dask/dask/issues/9135. I don't consider the example reported in https://github.com/dask/dask/issues/9135 sufficient motivation for this behavior but cloudpickle itself states that it is not deterministic

https://github.com/cloudpipe/cloudpickle/issues/453
https://github.com/cloudpipe/cloudpickle/issues/385

and I'm a bit on the fence about this since I do not know what to expect now.

Comment (fjetter) [reply dask/base.py:1222]: For instance, should this function be idempotent?

```python
import dask
from dask.base import tokenize
dask.config.set({"tokenize.ensure-deterministic": True})

def foo():
    import pickle
    class A:
        pass
    return tokenize(A())

assert foo() == foo()  # Should this be True or False?
```

For this PR this is `False` since every invocation of `foo` returns a different token. On `main` this is raising a `RuntimeError`

Comment (fjetter) [reply dask/base.py:1222]: The intention of the `RuntimeError` is to alert a user that a call includes a custom object that does not implement the `__dask_tokenize__` protocol, see https://github.com/dask/dask/issues/6555 for the first discussion

Comment (fjetter) [reply dask/base.py:1222]: For the record, this dynamic class determinism is not working with cloudpickle due to a uuid that is being used internally, see https://github.com/cloudpipe/cloudpickle/blob/d003266b18336e1e603536bdbe6518bc2dcc00d3/cloudpickle/cloudpickle.py#L112
i.e. cloudpickle is intentionally distinguishing the two calls. Do we want or need to do the same? From a dask user perspective, I think both calls are basically identical.

Maybe this doesn't matter, I don't know

Comment (fjetter) [reply dask/base.py:1222]: my gut tells me it is safer to raise if we are not sure. At this time it is hard to estimate the impact on dask-expr later and removing an exception later on is easier than introducing one

Comment (crusaderky) [reply dask/base.py:1222]: It's worth noting that your example fails for local _instances_, but passes for local _classes_ and _functions_:

```python
# PASS
def test_tokenize_local_classes_from_different_contexts():
    def f():
        class C:
            pass

        return C

    assert check_tokenize(f()) == check_tokenize(f())


# FAIL
def test_tokenize_local_instances_from_different_contexts():
    def f():
        class C:
            pass

        return C()

    assert check_tokenize(f()) == check_tokenize(f())


# PASS
def test_tokenize_local_functions_from_different_contexts():
    def f():
        def g():
            return 123

        return g

    assert check_tokenize(f()) == check_tokenize(f())

```

IMHO it's not terribly common to define whole classes in a local context and I can't think of a real-life scenario where this will make things break. I'm not saying impossible, just that for me it feels like an edge case.

I've added the three tests above and xfailed the instance one, linking the upstream cloudpickle issue.

Comment (crusaderky) [reply dask/base.py:1222]: Important to note is that the difference in token is generated by calling f() twice. If you call f() once, then pass the returned object through a cloudpickle round-trip, the token remains the same.

Comment (crusaderky) [reply dask/base.py:1222]: Also worth noting that `assert f() == f()` fails in all three above cases. So to me in the two cases that pass we're already doing more than what the Python interpreter does natively.
